jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java

```diff
@@ -0,0 +1,283 @@
+    {
+        if (nextIndex < a.size()) {
+            return true;
+        }
+        tryAddNext();
+        return nextIndex < a.size();
+    }
+
+    @Override
+    public GraphPath<V, E> next()
+    {
+        if (nextIndex < a.size()) {
+            return a.get(nextIndex++);
+        }
+        tryAddNext();
+        if (nextIndex >= a.size()) {
+            throw new NoSuchElementException();
+        }
+        return a.get(nextIndex++);
+    }
+
+    /**
+     * Determine and store in {@link #b} the next shortest paths and determine the actual shortest
+     * path and add it to {@link #a}. Instead of altering the underlying graph, this implementation
+     * masks edges and nodes that shall not be considered using {@link MaskSubgraph}.
+     */
+    private void tryAddNext()
+    {
+        GraphPath<V, E> lastShortestPath = a.get(a.size() - 1);
+        // The spur node ranges from the first node to the next to last
+        // node in the previous k-shortest path.
+        for (int i = 0; i < lastShortestPath.getLength() - 1; i++) {
+            // Spur node is retrieved from the previous k-shortest path
+            V spurNode = lastShortestPath.getVertexList().get(i);
+            // The sequence of nodes from the source to the spur node of
+            // the previous k-shortest path.
+            List<V> rootPath = lastShortestPath.getVertexList().subList(0, i);
+            List<V> maskedVertices = new ArrayList<V>();
+            List<E> maskedEdges = new ArrayList<E>();
+            for (GraphPath<V, E> path : a) {
+                if (rootPath.equals(path.getVertexList().subList(0, i))) {
+                    // Remove the links that are part of the previous
+                    // shortest paths which share the same root path.
+                    maskedEdges.add(
+                        graph
+                            .getEdge(path.getVertexList().get(i), path.getVertexList().get(i + 1)));
+                }
+            }
+            for (V rootPathNode : rootPath) {
+                if (!rootPathNode.equals(spurNode)) {
+                    maskedVertices.add(rootPathNode);
+                }
+            }
+
+            MaskSubgraph<V, E> maskedSubgraph;
+            if (graph instanceof DirectedGraph) {
+                maskedSubgraph = new DirectedMaskSubgraph<V, E>(
+                    (DirectedGraph<V, E>) graph, maskedVertices::contains, maskedEdges::contains);
+            } else {
+                maskedSubgraph =
+                    new MaskSubgraph<V, E>(graph, maskedVertices::contains, maskedEdges::contains);
+            }
+            // Calculate the spur path from the spur node to the sink.
+            ShortestPathAlgorithm<V, E> maskedShortestSpurPathAlgorithm =
+                spFactory.apply(maskedSubgraph);
+            GraphPath<V, E> spurPath = maskedShortestSpurPathAlgorithm.getPath(spurNode, sink);
+
+            if (null != spurPath && 0 != spurPath.getLength()) { // BellmannFord

```

Comment:
BellmanFord is (by contract) required to return null if no path exists. This is sounds like a bug. Could you indicate where this occurs?

Reply:
Just take a look at the result of org.jgrapht.alg.shortestpath.MultiShortestPathsStringDWETest
[YenKShortestPaths+BellmanFord/Eppstein_11-0]
BellmannFord returns an empty path.

Reply:
Opened #433 to resolve this. Adjust accordingly after merge.

Reply:
Removed the special handling of Bellman after #433 was merged.